libraries/botbuilder-azure/src/cosmosDbPartitionedStorage.ts

```diff
@@ -229,41 +242,76 @@ export class CosmosDbPartitionedStorage implements Storage {
                             this.cosmosDbStorageOptions.keySuffix,
                             this.cosmosDbStorageOptions.compatibilityMode
                         ),
-                        realId: k,
-                        document: changesCopy,
+                        realId: key,
+                        document: change,
                     });
 
-                    const eTag: string = changes[k].eTag;
-                    const ac =
+                    const accessCondition =
                         eTag !== '*' && eTag != null && eTag.length > 0
                             ? { accessCondition: { type: 'IfMatch', condition: eTag } }
                             : undefined;
+
                     try {
-                        await this.container.items.upsert(documentChange, ac);
+                        await this.container.items.upsert(document, accessCondition);
                     } catch (err) {
-                        this.throwInformativeError('Error upserting document', err);
+                        // This check could potentially be performed before even attempting to upsert the item
+                        // so that a request wouldn't be made to Cosmos if it's expected to fail.
+                        // However, performing the check here ensures that this custom exception is only thrown
+                        // if Cosmos returns an error first.
+                        // This way, the nesting limit is not imposed on the Bot Framework side
+                        // and no exception will be thrown if the limit is eventually changed on the Cosmos side.
+                        const nestingError = this.getNestingError(change);
+
+                        this.throwInformativeError(nestingError ?? 'Error upserting document', err);
                     }
                 }
             )
         );
     }
 
+    // Return an informative error message if upsert failed due to deeply nested data
+    private getNestingError(json: object): string | undefined {
+        const checkDepth = (obj: unknown, depth: number, isInDialogState: boolean) => {
+            if (depth > maxDepthAllowed) {
+                const message = `Maximum nesting depth of ${maxDepthAllowed} exceeded.`;
+
+                if (isInDialogState) {
+                    return (
+                        `${message} This is most likely caused by recursive component dialogs. ` +
+                        'Try reworking your dialog code to make sure it does not keep dialogs on the stack ' +
+                        "that it's not using. For example, consider using ReplaceDialogAsync instead of BeginDialogAsync."
+                    );
+                } else {
+                    return `${message} Please check your data for signs of unintended recursion.`;
+                }
+            } else if (obj && typeof (obj) === "object") {
+                depth++;
+                return Object.entries(obj).reduce(
```

Comment:
I'd write this using `find` instead of `reduce`. I think it's simpler to understand, and iteration will stop as soon as there is an error case.

Reply:
https://www.typescriptlang.org/play?#code/MYewdgzgLgBAtgQwB4BECmAHKALAggG3xAHc0ATGAXhgEYBuAKAdElmGzWAGt0tsqYAChAAjAFYAuGAFcwXMCTAAaGGUw4pYaXBFoATioCWEAJJgUhhEQDmAZSgIoaKSJAh8aBGACUU6HsMwaxgAHxkwNQAzQPIqAD4YAG8GGFSYQ0ihNT4YBMRUdTxCEnJvJJS0ypZoeDQICARrNAEAAwBZZEM4bRgwOqhA4OycGBBMgBJE-N4cAiJSMgBfGDQkYDRycgA6FsYKytSMoWMzCysQOwcnMuSDu5g9NChpPTAhffu7lsm4Ooam5YAFWwxnSEHgIBq+EMXDQ+AAnjBgAhpBBYiJEY9gC8IIYAG7NUBwDDgNBgWBkSw2CBbGAtGAAag+n0qAHJAXpMWhiCA9FxBjB4SAXqoqRckSA1DAoCB4AhYTAIC9moYKSA6r0QLBYZhRedrODwNKOIqHNwYKzGcyWakAEQ4RzpKCs8EKWCowa0gBivJWSAQxI8KmqhjUehkuKCMAASph8Ah1mcbLgIPCwMB0qxPBQxjAAEJoayBJMXFNp4BbW3Wu7eRifZZwtHlG2pR7PV50n5-RpoZYABQ8CCb7E4XEFwvDZEcCBgkV9uOskFGmVkgScEViWJxhnAOzr90WHwb+CbR2E4hgADJL9L4Rg0Lnz2IypRXzBbaIxJwoLabtWJawMAANqwvCKh4lY0hoAAugIADy4jflsZJQAEdRPt4WzRBEgj-pUgggWgYEwBB+BQdBL4JCO3AzNguEtgcpFQUoeH3MM-AMrQLEMZUoFUG+rKUvq9gJlwlohGEJzmGKlyOGg3hlAA-IpwHQXsnxti8bxMWg+6pIeHyaR2shRDEZCMIsezVO4aBbDYgjUTwhSCIkc4gFIiQiAgegeV5ABeUgAAyLCFKiBSokRWGiClAA

Reply:
EDIT: `reduce` makes more sense as we want the results of the recursive function, not the offending object. My mistake!